Stats step 1: stats v4 UI

jaclync opened the pull request:
Fixes #1067 
Based on https://github.com/woocommerce/woocommerce-ios/pull/1184

- [x] ⚠️ Update PR base to `develop` before merging ⚠️ 

## Changes
- Sorry about the size of this PR, most code are view controllers (and a xib from v3)
- Reset stats v4 data on switching store
- Added `chartAxisHourFormatter: DateFormatter` to show hour date format on stats v4 "today" chart x-axis
- The changes in `UIView+Helpers` are intended to be in `WordPressUI` and currently pending code review: https://github.com/wordpress-mobile/WordPressUI-iOS/pull/39
- In `DashboardUIFactory`, showed stats v4 UI if building from Xcode
- In `PeriodDataViewController` and `TopPerformerDataViewController`, moved protocol conformance to the extension that implements the `IndicatorInfoProvider` protocol
- (The post about stats v4 UI hierarchy would be helpful for reviewing the view controllers) Top container view controller `StoreStatsAndTopPerformersViewController`:
  - `StoreStatsAndTopPerformersViewController` is a subclass of `ButtonBarPagerTabStripViewController`, and a lot of button tab bar setup is the same as in v3 `StoreStatsViewController` and `TopPerformersViewController`
  - 4 child view controllers `StoreStatsAndTopPerformersViewController` for each time range tab:
    - `StoreStatsAndTopPerformersViewController` is a container of scrollable stack view of:
      - Store stats data view, managed by child view controller `StoreStatsV4PeriodViewController`
        - `StoreStatsV4PeriodViewController` is mostly the same as v3 `PeriodDataViewController` (and `PeriodDataViewController.xib`) with time range instead of granularity enum and different Fetched Results Controller setup for store stats and site visit stats
      - Top performers header view `TopPerformersSectionHeaderView` (created in this PR)
      - Top performers data view, managed by child view controller `TopPerformerDataViewController` (same in v3)

![Stats UI work](https://user-images.githubusercontent.com/1945542/63013089-08126e00-bebe-11e9-9cd5-82e405dabf4a.png)


## Testing
Prerequisite: the site has WC admin extension installed and activated
- Build this branch and run from Xcode for stats v4 UI
- Go to "My store" tab
- Check each time range tab: "Today"/"This Week"/"This Month"/"This Year" --> the stats should be consistent with the WC admin dashboard on `/wp-admin`

## Example screenshots

\ | Today | This week | This month | This year
-- | -- | -- | -- | --
Store stats | ![Simulator Screen Shot - iPhone 5s - 2019-08-15 at 17 19 49](https://user-images.githubusercontent.com/1945542/63086022-83892380-bf81-11e9-9407-d109bc598166.png) | ![Simulator Screen Shot - iPhone 5s - 2019-08-15 at 17 19 56](https://user-images.githubusercontent.com/1945542/63085975-6eac9000-bf81-11e9-8f30-347a45b245ca.png) | ![Simulator Screen Shot - iPhone 5s - 2019-08-15 at 17 20 03](https://user-images.githubusercontent.com/1945542/63085977-6eac9000-bf81-11e9-894e-90d25a3d950e.png) | ![Simulator Screen Shot - iPhone 5s - 2019-08-15 at 17 20 08](https://user-images.githubusercontent.com/1945542/63085978-6f452680-bf81-11e9-9e39-2444ee46d2a9.png)
Top performers | ![Simulator Screen Shot - iPhone 5s - 2019-08-16 at 11 37 48](https://user-images.githubusercontent.com/1945542/63142340-019a0880-c01c-11e9-9b4d-1fe12523704b.png) | ![Simulator Screen Shot - iPhone 5s - 2019-08-16 at 11 37 51](https://user-images.githubusercontent.com/1945542/63142344-08288000-c01c-11e9-8870-9109784f83c8.png) | ![Simulator Screen Shot - iPhone 5s - 2019-08-16 at 11 49 59](https://user-images.githubusercontent.com/1945542/63142347-0e1e6100-c01c-11e9-9d6d-8a6131882e0a.png) | ![Simulator Screen Shot - iPhone 5s - 2019-08-16 at 11 49 45](https://user-images.githubusercontent.com/1945542/63142353-11b1e800-c01c-11e9-941e-bb0c589e534d.png)









Update release notes:

- [ ] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

mindgraffiti commented:
I was reviewing the code and realized, I'm not sure what type of testing do you run? A store with v3 stats and a store with v4? Any other test cases I should try out?

jaclync commented:
> I was reviewing the code and realized, I'm not sure what type of testing do you run? A store with v3 stats and a store with v4? Any other test cases I should try out?

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersPeriodViewController.swift:
```diff
@@ -0,0 +1,215 @@
+        scrollView.pinSubviewToAllEdges(stackView)
+        NSLayoutConstraint.activate([
+            stackView.widthAnchor.constraint(equalTo: scrollView.widthAnchor)
+            ])
```
Closing parens tabbed over one too many

jaclync replied:
hmm this is how Xcode formats the indentation though, if you move that line around or ctrl+i the whole code block

jaclync replied:
maybe we can look into some auto-formatting tool like swift lint to detect/correct these formatting details for us?

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersPeriodViewController.swift:
```diff
@@ -0,0 +1,215 @@
+        stackView.addArrangedSubview(storeStatsPeriodView)
+        NSLayoutConstraint.activate([
+            storeStatsPeriodView.heightAnchor.constraint(equalToConstant: 380),
+            ])
```
same here

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersPeriodViewController.swift:
```diff
@@ -0,0 +1,215 @@
+            headerBottomBorderView.bottomAnchor.constraint(equalTo: topPerformersHeaderView.bottomAnchor),
+            headerBottomBorderView.leadingAnchor.constraint(equalTo: topPerformersHeaderView.leadingAnchor),
+            headerBottomBorderView.trailingAnchor.constraint(equalTo: topPerformersHeaderView.trailingAnchor),
+            ])
```
and here

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersPeriodViewController.swift:
```diff
@@ -0,0 +1,215 @@
+        NSLayoutConstraint.activate([
+            topPerformersPeriodView.heightAnchor.constraint(equalToConstant: 359.5),
+            topPerformersPeriodView.heightAnchor.constraint(greaterThanOrEqualToConstant: 359.5)
+            ])
```
ditto

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersPeriodViewController.swift:
```diff
@@ -0,0 +1,215 @@
+        let topPerformersPeriodView = topPerformersPeriodViewController.view!
+        stackView.addArrangedSubview(topPerformersPeriodView)
+        NSLayoutConstraint.activate([
+            topPerformersPeriodView.heightAnchor.constraint(equalToConstant: 359.5),
```
why `.5` in this constraint? Is this to account for the border line getting drawn in?

jaclync replied:
I calculated the height from the v3 top performers height of 465px, deducting the following UI components that are in the vertical stack view (in `Dashboard.storyboard`):

header view (moved to a separate view `TopPerformersSectionHeaderView`): 55.5px
border: 1px
button bar view (moved to the top as the time range selector): 48px
border: 1px
——
105.5px

Then I got 359.5px for just the top performers data view

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersPeriodViewController.swift:
```diff
@@ -0,0 +1,215 @@
+        view.backgroundColor = StyleManager.wooGreyBorder
+        NSLayoutConstraint.activate([
+            view.heightAnchor.constraint(equalToConstant: 1)
+            ])
```
ditto

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersViewController.swift:
```diff
@@ -0,0 +1,406 @@
+        view.backgroundColor = StyleManager.wooGreyBorder
+        NSLayoutConstraint.activate([
+            view.heightAnchor.constraint(equalToConstant: 1)
+            ])
```
Closing parens tabbed over one too many

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsV4PeriodViewController.swift:
```diff
@@ -0,0 +1,609 @@
+
+    /// Placeholder: Mockup Charts View
+    ///
+    private lazy var placehoderChartsView: ChartPlaceholderView = ChartPlaceholderView.instantiateFromNib()
```
```suggestion
    private lazy var placeholderChartsView: ChartPlaceholderView = ChartPlaceholderView.instantiateFromNib()
```

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsV4PeriodViewController.swift:
```diff
@@ -0,0 +1,609 @@
+            return lastUpdatedDate.relativelyFormattedUpdateString
+        } else {
+            return ""
+        }
```
This block can be re-written to remove the `else`:

```
if let lastUpdatedDate = lastUpdatedDate {
            return lastUpdatedDate.relativelyFormattedUpdateString
        } 

return ""
}
```

jaclync commented:
Addressed the top performers bottom border in https://github.com/woocommerce/woocommerce-ios/pull/1185/commits/455e1739ebd44d295878890e58d5757fe52266fe (also added a 44px empty view below it, the design spec has 181px spacing and I'll confirm with Kylea). Some screenshots:

no data view | 2 top performers
-- | --
![Simulator Screen Shot - iPhone 5s - 2019-08-16 at 11 37 48](https://user-images.githubusercontent.com/1945542/63142162-4f624100-c01b-11e9-88cf-4fd981ed1cf6.png) | ![Simulator Screen Shot - iPhone 5s - 2019-08-16 at 11 37 51](https://user-images.githubusercontent.com/1945542/63142163-4f624100-c01b-11e9-987e-d11ab9a3eeef.png)

jaclync commented:
Responded to the CR comments with some corresponding code updates in the last 3 commits. Merging now, lemme know if there is anything I missed and I'll address in a separate PR. Also let's revisit the [formatting topic](https://github.com/woocommerce/woocommerce-ios/pull/1185#discussion_r314569221) in p2!
